haskell-style.md

```diff
@@ -660,3 +660,329 @@ module Driver
     ) where
 
   ```
+
+## Haddocks
+
+Haskell modules should be commented with valid Haddock documentation. We are not
+yet requiring a certain level of coverage, but it is strongly encouraged.
+
+### General guides
+
+- Use proper Haddock [markup][]
+- Link all identifiers, anywhere they appear
+
+  Bad
+
+  ```hs
+  -- | Construct a @'FlipFlop'@
+  --
+  -- If the size is right, you will get a @Right FlipFlop@, otherwise a @Left@
+  --
+  ```
+
+  Good
+
+  ```hs
+  -- | Construct a @'FlipFlop'@
+  --
+  -- If the size is right, you will get a @'Right' 'FlipFlop'@, otherwise a
+  -- @'Left'@.
+  --
+  ```
+
+- Use leading documentation (`-- |`) for top-level definitions and trailing
+  documentation (`-- ^`) for record attributes and function arguments
+
+  Bad
+
+  ```hs
+  data Foo = Foo
+    {
+    -- | Foo's foo
+      fooFoo :: Foo
+    -- | Foo's bar
+    , fooBar :: Bar
+    }
+  -- ^ What even is a Foo?
+  ```
+
+  Good
+
+  ```hs
+  -- | What even is a Foo?
+  data Foo = Foo
+    { fooFoo :: Foo
+    -- ^ Foo's foo
+    , fooBar :: Bar
+    -- ^ Foo's bar
+    }
+  ```
+
+  Best, for this case
+
+  ```hs
+  -- | What even is a Foo?
+  data Foo = Foo
+    { fooFoo :: Foo -- ^ Foo's foo
+    , fooBar :: Bar -- ^ Foo's bar
+    }
+  ```
+
+  **NOTE**: do not align trailing documentation in context-sensitive ways.
+
+  Bad
+
+  ```hs
+  -- | What even is a Foo?
+  data Foo = Foo
+    { fooFoos :: [Foo] -- ^ Foo's foos
+    , fooBar :: Bar    -- ^ Foo's bar
```

Comment:
I would add a really long comment here too for the Bad case

```haskell
  , fooBaz :: Baz -- ^ This is a really long comment and we should make sure we
                  -- don't align this in context-sensitive ways
```

Up to you

Reply:
Oh yeah! I wanted to mention that the usual Summary/Body guides apply to trailing comments. And that trailing comments can only be on the same line if they're short enough.

So:

```hs
, fooBaz :: Baz
-- ^ The Baz of this foo
--
-- And  this is what you do when it gets really long and needs a Body.
---
```

If I don't get immediate push back on this comment, I'll add that to the PR.